### PR TITLE
Clean up code, use DiagnosticInfo in LibraryImportGenerator, Cache well known attribute types in StubEnvironment

### PR DESCRIPTION
**Pull request by jtschuster**
Remove unused usings and simplify access to CodeAnalysis.Location.

Cache the attributes searched for in incremental stub calculations.

Add warning if unsafe blocks are not allowed.

Removes assert in ComMethodInfo that was throwing when editing files in the test project.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx [hunk 1]
```diff
@@ -267,6 +267,15 @@
   <data name="InvalidStringMarshallingConfigurationOnInterfaceMessage" xml:space="preserve">
     <value>The configuration of 'StringMarshalling' and 'StringMarshallingCustomType' on interface '{0}' is invalid. {1}</value>
   </data>
+  <data name="RequiresAllowUnsafeBlocksDescription" xml:space="preserve">
+    <value>LibraryImportAttribute requires unsafe code. Project must be updated with '&lt;AllowUnsafeBlocks&gt;true&lt;/AllowUnsafeBlocks&gt;'.</value>
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx:271]:
```suggestion
    <value>GeneratedComInterfaceAttribute requires unsafe code. Project must be updated with '&lt;AllowUnsafeBlocks&gt;true&lt;/AllowUnsafeBlocks&gt;'.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx [hunk 2]
```diff
@@ -267,6 +267,15 @@
...
+  </data>
+  <data name="RequiresAllowUnsafeBlocksMessage" xml:space="preserve">
+    <value>LibraryImportAttribute requires unsafe code. Project must be updated with '&lt;AllowUnsafeBlocks&gt;true&lt;/AllowUnsafeBlocks&gt;'.</value>
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx:274]:
```suggestion
    <value>GeneratedComInterfaceAttribute requires unsafe code. Project must be updated with '&lt;AllowUnsafeBlocks&gt;true&lt;/AllowUnsafeBlocks&gt;'.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx [hunk 3]
```diff
@@ -267,6 +267,15 @@
...
+  </data>
+  <data name="RequiresAllowUnsafeBlocksTitle" xml:space="preserve">
+    <value>LibraryImportAttribute requires unsafe code.</value>
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx:277]:
```suggestion
    <value>GeneratedComInterfaceAttribute requires unsafe code.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx (hunk 1 above)
Comment (jtschuster) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx:271]: Oops, thanks for catching that.

**Review by jkoritzinsky — APPROVED** [review]
LGTM once the build is fixed.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs [hunk 4]
```diff
@@ -26,8 +25,10 @@ internal sealed record ComInterfaceInfo(
...
         Location DiagnosticLocation)
     {
-        public static DiagnosticOrInterfaceInfo From(INamedTypeSymbol symbol, InterfaceDeclarationSyntax syntax, CancellationToken _)
+        public static DiagnosticOrInterfaceInfo From(INamedTypeSymbol symbol, InterfaceDeclarationSyntax syntax, StubEnvironment env, CancellationToken _)
         {
+            if (env.Compilation.Options is not CSharpCompilationOptions { AllowUnsafe: true }) // Unsafe code enabled
+                return DiagnosticOrInterfaceInfo.From(DiagnosticInfo.Create(GeneratorDiagnostics.RequiresAllowUnsafeBlocks, syntax.Identifier.GetLocation()));
             // Verify the method has no generic types or defined implementation
             // and is not marked static or sealed
             if (syntax.TypeParameterList is not null)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs:34]: Is this a bug? We are checking this here and the same thing within it? Should one of these be `symbol`?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs:34]: @jkoritzinsky ?

Comment (jtschuster) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs:34]: Yeah, it should be just one check

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs [hunk 5]
```diff
@@ -52,18 +52,14 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             // Validate if attributed methods can have source generated
             var methodsWithDiagnostics = attributedMethods.Select(static (data, ct) =>
             {
-                Diagnostic? diagnostic = GetDiagnosticIfInvalidMethodForGeneration(data.Syntax, data.Symbol);
-                return new { Syntax = data.Syntax, Symbol = data.Symbol, Diagnostic = diagnostic };
+                DiagnosticInfo? diagnostic = GetDiagnosticIfInvalidMethodForGeneration(data.Syntax, data.Symbol);
+                if (diagnostic is not null)
+                    return DiagnosticOr<(MethodDeclarationSyntax Syntax, IMethodSymbol Symbol)>.From(diagnostic);
+                else
+                    return DiagnosticOr<(MethodDeclarationSyntax Syntax, IMethodSymbol Symbol)>.From((data.Syntax, data.Symbol));
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:59]: If there is more than one predicate or more than one clause (i.e., either `else if` or `else`) then I prefer to have braces. You can also just use a terinary operator here if you like.

**Pull request merged**